Serve OCaml Compiler Manuals

mtelvers opened the pull request:
* Include the precompiled html manuals from [ocaml-web/html-compiler-manuals](https://github.comocaml-web/html-compiler-manuals) into the `Dockerfile`.  The `RUN` line determines which branches exist and clones them all into local directories: `/manual/4.14`, `/manual/5.0`, etc; updating this is unnecessary when a new version is released.
* Updated the code to serve `/manual` and below from the local filesystem
* Added default name `index.html` when the directory is requested
* Updated the default manual link `/manual` to serve from these files rather than redirect to v2.ocaml.org

mtelvers committed:
Added local serving of HTML manuals

mtelvers committed:
Update format

shakthimaan commented on src/ocamlorg_web/lib/redirection.ml:
```diff
@@ -449,6 +449,25 @@ let from_v2 =
 
+let default_index_html =
+  [
+    ("/manual", Url.manual_lang latest_version);
```
We will need to add one for `latest`.

sabine replied:
We don't need a redirect for `latest`, we need to serve this via Dream under `/manual/latest`. A redirect would mean that the URL in the browser would change to the specific version.

sabine commented on src/ocamlorg_web/lib/redirection.ml:
```diff
@@ -459,12 +478,13 @@ let redirect_p pattern =
+
 let manual =
   [
+    serve "/manual/**" "/manual/";
```
I don't think we should be configuring pages to be served from the `redirection.ml` module. Instead, this should be put into `router.ml`.

mtelvers replied:
I have moved this to `router.ml`

sabine commented on src/global/url.ml:
```diff
@@ -46,7 +46,9 @@ let minor v =
-let manual = "/releases/latest/manual.html"
+let manual = "/manual"
+let manual_lang v = "/manual/" ^ minor v ^ "/index.html"
+let manual_api v = "/manual/" ^ minor v ^ "/api/index.html"
```
`index.html` is the default page looked up by the browser, so we don't need to append this here. Shorter URLs tend to be better. I would rather redirect the other way.

mtelvers replied:
The browser doesn't request the default page.  If configured, the server returns a default page when the browser doesn't specify a specific page.

shakthimaan replied:
When manually testing instructions given in the Dockerfile, I am getting a 404 for /manual/5.1/index.html. What could I be missing?

mtelvers replied:
Are your manuals cloned to the top level root, aka /, of your test machine?

cuihtlauac replied:
Could we have a setup with an environment variable (maybe `OCAMLORG_MANUAL_PATH` ?) such as in `src/ocamlorg_web_lib/config.ml`? That would be easier for debugging.

shakthimaan replied:
> Are your manuals cloned to the top level root, aka /, of your test machine?

Ohh! Okay. After copying to /manuals, it works fine. Thanks!

cuihtlauac replied:
This branch: https://github.com/ocaml/ocaml.org/tree/html-compiler-manuals-config contains a patch that allows to set the manual path in a shell environment variable

mtelvers replied:
Thanks, I'm on leave this week.  I have cherry-picked your commit to this PR.

cuihtlauac commented:
Using the following on the upstream `https://github.com/ocaml-web`:
```bash
git grep -ho 'href=\"https://[._a-z0-9/\-]*' | sort -u
```

I found the following hardcoded urls:
```
https://ocaml.org/
https://ocaml.org/releases/
https://ocaml.org/releases/4.13/ocaml-4.13-refman-html.tar.gz
https://ocaml.org/releases/4.13/ocaml-4.13-refman.info.tar.gz
https://ocaml.org/releases/4.13/ocaml-4.13-refman.pdf
https://ocaml.org/releases/4.13/ocaml-4.13-refman.txt
https://ocaml.org/releases/4.14/ocaml-4.14-refman-html.tar.gz
https://ocaml.org/releases/4.14/ocaml-4.14-refman.info.tar.gz
https://ocaml.org/releases/4.14/ocaml-4.14-refman.pdf
https://ocaml.org/releases/4.14/ocaml-4.14-refman.txt
https://ocaml.org/releases/5.0/ocaml-5.0-refman-html.tar.gz
https://ocaml.org/releases/5.0/ocaml-5.0-refman.info.tar.gz
https://ocaml.org/releases/5.0/ocaml-5.0-refman.pdf
https://ocaml.org/releases/5.0/ocaml-5.0-refman.txt
https://ocaml.org/releases/5.1/ocaml-5.1-refman-html.tar.gz
https://ocaml.org/releases/5.1/ocaml-5.1-refman.info.tar.gz
https://ocaml.org/releases/5.1/ocaml-5.1-refman.pdf
https://ocaml.org/releases/5.1/ocaml-5.1-refman.txt
https://ocaml.org/releases/5.2/ocaml-5.2-refman-html.tar.gz
https://ocaml.org/releases/5.2/ocaml-5.2-refman.info.tar.gz
https://ocaml.org/releases/5.2/ocaml-5.2-refman.pdf
https://ocaml.org/releases/5.2/ocaml-5.2-refman.txt
https://ocaml.org/releases/5.3/ocaml-5.3-refman-html.tar.gz
https://ocaml.org/releases/5.3/ocaml-5.3-refman.info.tar.gz
https://ocaml.org/releases/5.3/ocaml-5.3-refman.pdf
https://ocaml.org/releases/5.3/ocaml-5.3-refman.txt
```

All but the first two trampolines to v2. We should all turn them into relative links.

cuihtlauac commented on Dockerfile:
```diff
@@ -34,6 +34,8 @@ COPY --from=build /home/opam/_build/default/src/ocamlorg_web/bin/main.exe /bin/s
 
 COPY playground/asset playground/asset
 
+RUN git clone https://github.com/ocaml-web/html-compiler-manuals /manual && for a in $(git -C /manual branch -r --format='%(refname:lstrip=3)') ; do git -C /manual submodule add ./ $a && git -C /manual/$a checkout $a; done
```
This seems consistent with the `README.md` in `https://github.com/ocaml-web/html-compiler-manuals`. It says:
> For every compiler version, there is a branch named like exactly like the compiler version. The branch contains the HTML pages for the compiler manual to be served by OCaml.org.

However, I only see a single branch upstream. Am I missing something or can we do the `git clone` without iterating over branches?

shakthimaan replied:
The requirements have changed. We will need to update the README to reflect the same. Thanks for your review!

mtelvers replied:
With the new repo layout the `RUN` line can be greatly simplified.

cuihtlauac commented on src/global/url.ml:
```diff
@@ -46,7 +46,9 @@ let minor v =
+let manual_lang v = "/manual/" ^ minor v ^ "/index.html"
+let manual_api v = "/manual/" ^ minor v ^ "/api/index.html"
 let api_with_version v = v2 ^ "/releases/" ^ minor v ^ "/api/index.html"
 let api = "/releases/latest/api/index.html"
```
```suggestion
let api_with_version v = "/releases/" ^ minor v ^ "/api"
let api = "/releases/latest/api"
```

mtelvers replied:
I agree with this change, but I think it's outside the scope of this PR.  A whole exercise will need to be carried out to remove redirects to `v2`.  See `fwd_v2` in `src/ocamlorg_web/lib/redirection.ml`.

cuihtlauac replied:
I agree removing obsolete redirects should be in another PR. But this only allows the links to the Standard Library API (such as found in https://ocaml.org/docs, the footer and several other places) to point to the local content instead of V2.

mtelvers committed:
Updates following review

Commit:
Set manual path in config variable

mtelvers committed:
Reformat cherry picked commit

Commit:
Remove

* Repair broken API links to v2
* Deduplicate
  - Url.manual_lang into Url.manual_with_version
  - Url.manual_api into Url.api_with_version

cuihtlauac commented:
The changes I'm proposing to handle api urls are in this branch: https://github.com/ocaml/ocaml.org/tree/html-compiler-manuals-api

sabine commented:
@cuihtlauac I don't see a reason why we should introduce a config variable `OCAMLORG_MANUAL_PATH` for the manual path. The rest looks useful on first glance, but haven't checked in detail.

cuihtlauac commented:
> @cuihtlauac I don't see a reason why we should introduce a config variable `OCAMLORG_MANUAL_PATH` for the manual path. The rest looks useful on first glance, but haven't checked in detail.

That allows developers to put their local copy of `html-compiler-manual` wherever they want/need (not only `/`).

mtelvers committed:
Latest updates

mtelvers commented:
I have added @cuihtlauac latest commit to this PR and included a commit to update line 50 of `src/global/url.ml` as the refactoring of these lines has made `api_with_version` the same as `manual_with_version` where they should be different:

```
-let api_with_version v = "/manual/" ^ minor v ^ "/index.html"
+let api_with_version v = "/manual/" ^ minor v ^ "/api/index.html"
```

@shakthimaan I note that `search_icon.svg` is missing from the `api` folder for all the versions of the manuals except for 4.12 and 4.13.  I think it needs to be included.

shakthimaan commented:
> search_icon.svg is missing from the api folder

The search_icon.svg has been added to the respective api/ folders.

sabine reviewed (approved):
LGTM, on staging this seems to work fine

Octachron commented:
Thanks a lot for the hard work on integrating the reference manual !
